crypto/bls_core.c

```diff
@@ -142,3 +141,147 @@ int bls_verify(const ep2_t pk, const byte* sig, const byte* data, const int len)
     else 
         return INVALID;
 }
+
+// Verifies a BLS signature in a byte buffer.
+// membership check of the signature in G1 is verified in this function
+// membership check of pk in G2 is not verified in this function
+// the membership check is separated to allow optimizing multiple verifications using the same key.
+int bls_verify(const ep2_t pk, const byte* sig, const byte* data, const int len) {  
+    ep_t s;
+    ep_new(s);
+    if (ep_read_bin_compact(s, sig, SIGNATURE_LEN) != RLC_OK) 
+        return INVALID;
+    
+    return bls_verify_ep(pk, s, data, len);
+}
+
+// binary tree structure to be used by bls_batch verify.
+// Each node contains a signature and a public key, the signature (resp. the public key) 
+// being the aggregated signature of the two children's signature (resp. public keys).
+// The leaves contain the initial signatures and public keys.
+typedef struct st_node { 
+    ep_st* sig;
+    ep2_st* pk;  
+    struct st_node* left; 
+    struct st_node* right; 
+} node;
+
+static node* new_node(const ep2_st* pk, const ep_st* sig){
+    node* t = (node*) malloc(sizeof(node));
+    t->pk = (ep2_st*)pk;
+    t->sig = (ep_st*)sig;
+    t->right = t->left = NULL;
+    return t;
+}
+
+// builds a binary tree of aggregation of signatures and public keys recursivley.
+static node* build_tree(const int len, const ep2_st* pks, const ep_st* sigs) {
+    // check if a leave is reached
+    if (len == 1) {
+        return new_node(pks, sigs);  // use the first element of the arrays
+    }
+
+    // a leave is not reached yet, 
+    int right_len = len/2;
+    int left_len = len - right_len;
+
+    // create a new node with new points
+    node* t = new_node((ep2_st*)malloc(sizeof(ep2_st)), (ep_st*)malloc(sizeof(ep_st)));
+    ep_new(t->sig);
+    ep_new(t->pk);
+    // build the tree in a top-down way
+    t->left = build_tree(left_len, &pks[0], &sigs[0]);
+    t->right = build_tree(right_len, &pks[left_len], &sigs[left_len]);
+    // sum the children
+    ep_add_projc(t->sig, t->left->sig, t->right->sig);
+    ep2_add_projc(t->pk, t->left->pk, t->right->pk); 
+    return t;
+}
+
+static void free_tree(node* root) {
+    // relic free
+    ep_free(root->sig);
+    ep2_free(root->pk);
+    if (root->left) {
```

Comment:
don't need to check right?

Reply:
Good comment, the tree is built so that a node either has 2 children or has no children at all.